swaylock/render.c

```diff
@@ -0,0 +1,146 @@
+			cairo_fill_preserve(cairo);
+			cairo_set_source_rgb(cairo, 51.0 / 255, 0, 250.0 / 255);
+			cairo_stroke(cairo);
+		} break;
+		case AUTH_STATE_INVALID: {
+			cairo_set_source_rgba(cairo, 250.0 / 255, 0, 0, 0.75);
+			cairo_fill_preserve(cairo);
+			cairo_set_source_rgb(cairo, 125.0 / 255, 51.0 / 255, 0);
+			cairo_stroke(cairo);
+		} break;
+		default: break;
+		}
+
+		// Draw a message
+		char *text = NULL;
+		cairo_set_source_rgb(cairo, 0, 0, 0);
+		cairo_select_font_face(cairo, "sans-serif",
+				CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+		cairo_set_font_size(cairo, ARC_RADIUS / 3.0f);
+		switch (state->auth_state) {
+		case AUTH_STATE_VALIDATING:
+			text = "verifying";
+			break;
+		case AUTH_STATE_INVALID:
+			text = "wrong";
+			break;
+		default: break;
+		}
+
+		if (text) {
+			cairo_text_extents_t extents;
+			double x, y;
+			cairo_text_extents(cairo, text, &extents);
+			x = (buffer_width / 2) -
+				(extents.width / 2 + extents.x_bearing);
+			y = (buffer_height / 2) -
+				(extents.height / 2 + extents.y_bearing);
+
+			cairo_move_to(cairo, x, y);
+			cairo_show_text(cairo, text);
+			cairo_close_path(cairo);
+			cairo_new_sub_path(cairo);
+		}
+
+		// Typing indicator: Highlight random part on keypress
+		if (state->auth_state == AUTH_STATE_INPUT
+				|| state->auth_state == AUTH_STATE_BACKSPACE) {
+			static double highlight_start = 0;
+			highlight_start +=
+				(rand() % (int)(M_PI * 100)) / 100.0 + M_PI * 0.5;
+			cairo_arc(cairo, buffer_width / 2, buffer_height / 2,
+					ARC_RADIUS, highlight_start,
+					highlight_start + TYPE_INDICATOR_RANGE);
+			if (state->auth_state == AUTH_STATE_INPUT) {
+				cairo_set_source_rgb(cairo, 51.0 / 255, 219.0 / 255, 0);
+			} else {
+				cairo_set_source_rgb(cairo, 219.0 / 255, 51.0 / 255, 0);
+			}
+			cairo_stroke(cairo);
+
+			// Draw borders
+			cairo_set_source_rgb(cairo, 0, 0, 0);
+			cairo_arc(cairo, buffer_width / 2, buffer_height / 2,
+					ARC_RADIUS, highlight_start,
+					highlight_start + TYPE_INDICATOR_BORDER_THICKNESS);
+			cairo_stroke(cairo);
+
+			cairo_arc(cairo, buffer_width / 2, buffer_height / 2,
+					ARC_RADIUS, highlight_start + TYPE_INDICATOR_RANGE,
+					highlight_start + TYPE_INDICATOR_RANGE +
+						TYPE_INDICATOR_BORDER_THICKNESS);
+			cairo_stroke(cairo);
+		}
+
+		// Draw inner + outer border of the circle
+		cairo_set_source_rgb(cairo, 0, 0, 0);
+		cairo_set_line_width(cairo, 2.0 * surface->scale);
+		cairo_arc(cairo, buffer_width / 2, buffer_height / 2,
+				ARC_RADIUS - ARC_THICKNESS / 2, 0, 2 * M_PI);
+		cairo_stroke(cairo);
+		cairo_arc(cairo, buffer_width / 2, buffer_height / 2,
+				ARC_RADIUS + ARC_THICKNESS / 2, 0, 2 * M_PI);
+		cairo_stroke(cairo);
+	}
+
+	wl_surface_set_buffer_scale(surface->surface, surface->scale);
+	wl_surface_attach(surface->surface, surface->current_buffer->buffer, 0, 0);
+	wl_surface_damage(surface->surface, 0, 0, buffer_width, buffer_height);
```

Comment:
If you want to damage `buffer_{width,height}`, use `wl_surface_damage_buffer`. Or you could just use `surface->{width,height}`.